### DISCUSSION TITLE
List/export groups of an organization

### DISCUSSION THREAD

**Discussion by darklex55** [dani-garcia/vaultwarden#4235] [Q&A]
Hi community. I've been trying to create a process for granting/revoking access to collections through a workflow tool by exposing and utilizing the vault api from the cli (bw serve). I have also deployed the connector (bwdc) to sync users and groups from Azure AD to an organization on bitwarden.

Some context to the process:

-> Groups are provisioned in Azure AD to both the workflow tool and bitwarden.
-> Access to collections in Bitwarden is granted through group membership (i.e. adding/removing people from groups) - this is handled through the Azure AD.
-> The workflow tool can change the group membership on Azure AD through the Graph API and in doing so, the provisioning will handle the change on Bitwarden.

**BUT there is a critical issue on the process above:** 

The workflow tool doesn't know the group mapping. I.e. when the tool is listing the available collections and is receiving as a response the list of groups that have access to each collection, the groups are referenced by the bitwarden's group id not the external id that is used for the group sync and provisioning with the connector. Since getting the full list of groups and their details is handled by the public API which is not currently supported by vaultwarden, does anyone have any idea on how to get a list/mapping of the internal group ids and their external ids through an automated export or an api?

**TLDR:** Is there a suggestion to access/export the list of groups defined in an organization programmatically, since the Organization (public) API is not supported?

**Comment by darklex55** [comment]
I'm answering my own question with a workaround I found. There is a repo [python-vaultwarden] (https://github.com/numberly/python-vaultwarden) that provides a python package of the bitwarden cli that is enhanced - i.e. including operations available normally from the public api and is compatible with vaultwarden. The drawback with that is that it does not come with a "serve" command, so it cannot be exposed as a REST API.

I've created a small flask api that is wrapping the python cli and I've enabled us to retrieve and alter the registered users. I'll leave this open if anyone suggests anything easier and better.

**Comment by peterge-misoft** [comment]
Here is the solution I found for exporting an organization using [Password Manager CLI](https://bitwarden.com/help/cli/#export):
`peterge@thinkpadt14s:~/Downloads/bw-linux-2024.3.1$ ./bw export --output /home/peterge/Documents/ --format json --organizationid 4ad08680-93e8-400e-a74b-cde5jf9842e3`

we came across this error, but the reason was that my account wasnt owner of the organization:
```
/snapshot/clients/apps/cli/build/bw.js:44022
                error.message != null
                      ^
TypeError: Cannot read properties of null (reading 'message')
    at Response.error (/snapshot/clients/apps/cli/build/bw.js:44022:23)
    at ExportCommand.<anonymous> (/snapshot/clients/apps/cli/build/bw.js:49408:33)
    at Generator.throw (<anonymous>)
    at rejected (/snapshot/clients/apps/cli/build/bw.js:49355:65)

Node.js v18.5.0
```
